lambda/src/server/slack.cljs

```diff
@@ -21,62 +26,133 @@
+                                                  (str "?" (uri/map->query-string query))))
+                    #js{:method "post"
+                        :Content-Type "application/json; charset=utf-8"
+                        :headers #js{:Authorization (str "Bearer " token)}
+                        :body (clj->js body) #_(.stringify js/JSON (clj->js body))}))
 
 (tasks/register-var! `post+)
 
-(defn post-query-string+ [family-method query-params]
-  ;; This fn is a hack to work around broken JSON bodies in Slack's API
-  (http/fetch-json+ (str base-uri family-method "?" (uri/map->query-string query-params))
-                    (j/lit {:method "post"
-                            :Content-Type "application/json; charset=utf-8"
-                            :headers {:Authorization (str "Bearer " bot-token)}})))
-
-(tasks/register-var! `post-query-string+)
-
 (comment
   (p/-> (get+ "users.list")
         println))
 
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 ;;;; Convenience wrappers over individual endpoints
-(defn views-open! [trigger-id blocks]
+(defn views-open! [token trigger-id blocks]
   (println "[views-open!] JSON blocks:" (blocks/to-json blocks))
-  (p/->> (post-query-string+ "views.open"
-                             {:trigger_id trigger-id
-                              :view (blocks/to-json blocks)})
+  (p/->> (post+ "views.open"
+                {:query {:trigger_id trigger-id
+                         :view (blocks/to-json blocks)}
+                 :token token})
          (http/assert-ok)
          ;; TODO better callback
          (println "slack views.open response:")))
 
 (tasks/register-var! `views-open!)
 
-(defn views-update! [view-id blocks]
-  (p/->> (post-query-string+ "views.update"
-                             {:view_id view-id
-                              :view (blocks/to-json blocks)})
+(defn views-update! [token view-id blocks]
+  (p/->> (post+ "views.update"
+                {:query {:view_id view-id
+                         :view (blocks/to-json blocks)}
+                 :token token})
          ;; TODO better callback
          (println "slack views.update response:")))
 
 (tasks/register-var! `views-update!)
+
+
+(defn slack-user-by-email
+  "Returns slack user id for the given email address, if found"
+  [email]
+  (p/let [res (get+ "users.lookupByEmail" {:query {:email email}})]
+    ;; UNCLEAR: how does Slack know what workspace we are interested in?
+    (when (j/get res :ok)
+      ;; user contains team_id, id, is_admin
+      (j/get res :user))))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;;; Oauth flow
+
+(j/defn req-origin [^:js {:keys [headers url query]
+                          {:keys [encrypted]} :connection}]
+  (let [host (j/get headers :host)]
+    (str "https://" host (str/replace url #"/slack.*" ""))))
+
+(j/defn install-redirect [^:js {:as req :keys [query]} res next]
```

Comment:
I may not fully understand when this gets invoked. An admin user installs the app, and then a separate rank-and-file user uses the global shortcut, and on their first time they get sent here? What directs them to the `"/slack/install"` route?

Reply:
I've added some documentation - https://github.com/sparkboard/sparkboard/pull/23/commits/4ce58ca49879e409804f8cccaf1917780d2fc50c